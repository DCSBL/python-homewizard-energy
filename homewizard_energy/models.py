"""Common models for HomeWizard Energy API."""

from __future__ import annotations

from dataclasses import asdict, dataclass, field
from datetime import datetime
from enum import Enum
from typing import Any

from mashumaro.config import BaseConfig
from mashumaro.exceptions import MissingField
from mashumaro.mixins.orjson import DataClassORJSONMixin


class BaseModel(DataClassORJSONMixin):
    """Base model for all HomeWizard models."""

    # pylint: disable-next=too-few-public-methods
    class Config(BaseConfig):
        """Mashumaro configuration."""

        serialize_by_alias = True
        omit_none = True


MODELS = {
    "HWE-P1": "p1dongle",
    "HWE-SKT": "energysocket",
    "HWE-WTR": "watermeter",
    "HWE-DSP": "display",
    "HWE-KWH1": "energymeter",
    "SDM230-wifi": "energymeter",
    "HWE-KWH3": "energymeter",
    "SDM630-wifi": "energymeter",
    "HWE-BAT": "battery",
}


def get_verification_hostname(model: str, serial_number: str) -> str:
    """Helper method to convert device model and serial to identifier

    The identifier is used to verify the device in the HomeWizard Energy API via HTTPS.
    """

    if model not in MODELS:
        raise ValueError(f"Unsupported model: {model}")

    return f"appliance/{MODELS[model]}/{serial_number}"


@dataclass(kw_only=True)
class Device(BaseModel):
    """Represent Device config."""

    product: Product | None = None
    id: str | None = None

    product_name: str = field()
    product_type: str = field()
    serial: str = field()
    api_version: str = field()
    firmware_version: str = field()

    @classmethod
    def __post_deserialize__(cls, obj: Device) -> Device:
        """Post deserialize hook for Device object."""
        _ = cls  # Unused

        obj.product = Product.from_type(obj.product_type)
        obj.id = get_verification_hostname(obj.product_type, obj.serial)
        return obj


@dataclass(kw_only=True)
class Measurement(BaseModel):
    """Represent Measurement."""

    # Deprecated, use 'System'
    wifi_ssid: str | None = field(
        default=None,
    )
    wifi_strength: int | None = field(
        default=None,
    )

    # Generic
    energy_import_kwh: float | None = field(
        default=None,
    )
    energy_import_t1_kwh: float | None = field(
        default=None,
    )
    energy_import_t2_kwh: float | None = field(
        default=None,
    )
    energy_import_t3_kwh: float | None = field(
        default=None,
    )
    energy_import_t4_kwh: float | None = field(
        default=None,
    )
    energy_export_kwh: float | None = field(
        default=None,
    )
    energy_export_t1_kwh: float | None = field(
        default=None,
    )
    energy_export_t2_kwh: float | None = field(
        default=None,
    )
    energy_export_t3_kwh: float | None = field(
        default=None,
    )
    energy_export_t4_kwh: float | None = field(
        default=None,
    )

    power_w: float | None = field(
        default=None,
    )
    power_l1_w: float | None = field(
        default=None,
    )
    power_l2_w: float | None = field(
        default=None,
    )
    power_l3_w: float | None = field(
        default=None,
    )

    voltage_v: float | None = field(
        default=None,
    )
    voltage_l1_v: float | None = field(
        default=None,
    )
    voltage_l2_v: float | None = field(
        default=None,
    )
    voltage_l3_v: float | None = field(
        default=None,
    )

    current_a: float | None = field(
        default=None,
    )
    current_l1_a: float | None = field(
        default=None,
    )
    current_l2_a: float | None = field(
        default=None,
    )
    current_l3_a: float | None = field(
        default=None,
    )

    apparent_power_va: float | None = field(
        default=None,
    )
    apparent_power_l1_va: float | None = field(
        default=None,
    )
    apparent_power_l2_va: float | None = field(
        default=None,
    )
    apparent_power_l3_va: float | None = field(
        default=None,
    )

    reactive_power_var: float | None = field(
        default=None,
    )
    reactive_power_l1_var: float | None = field(
        default=None,
    )
    reactive_power_l2_var: float | None = field(
        default=None,
    )
    reactive_power_l3_var: float | None = field(
        default=None,
    )

    power_factor: float | None = field(
        default=None,
    )
    power_factor_l1: float | None = field(
        default=None,
    )
    power_factor_l2: float | None = field(
        default=None,
    )
    power_factor_l3: float | None = field(
        default=None,
    )

    frequency_hz: float | None = field(
        default=None,
    )

    # P1 Meter specific
    timestamp: datetime | None = field(
        default=None,
        metadata={"deserialize": lambda x: Measurement.to_datetime(x)},
    )
    protocol_version: int | None = field(
        default=None,
    )
    meter_model: str | None = field(
        default=None,
    )
    unique_id: str | None = field(
        default=None,
        metadata={"deserialize": lambda x: Measurement.hex_to_readable(x)},
    )

    tariff: int | None = field(
        default=None,
    )

    voltage_sag_l1_count: int | None = field(
        default=None,
    )
    voltage_sag_l2_count: int | None = field(
        default=None,
    )
    voltage_sag_l3_count: int | None = field(
        default=None,
    )

    voltage_swell_l1_count: int | None = field(
        default=None,
    )
    voltage_swell_l2_count: int | None = field(
        default=None,
    )
    voltage_swell_l3_count: int | None = field(
        default=None,
    )

    any_power_fail_count: int | None = field(
        default=None,
    )
    long_power_fail_count: int | None = field(
        default=None,
    )

    average_power_15m_w: float | None = field(
        default=None,
    )
    monthly_power_peak_w: float | None = field(
        default=None,
    )
    monthly_power_peak_timestamp: datetime | None = field(
        default=None,
        metadata={"deserialize": lambda x: Measurement.to_datetime(x)},
    )

    external_devices: dict[str, ExternalDevice] | None = field(
        default=None,
        metadata={
            "alias": "external",
            "deserialize": lambda list: Measurement.array_to_external_device_list(list),
        },
    )

    # Watermeter Specific
    active_liter_lpm: float | None = field(
        default=None,
    )
    total_liter_m3: float | None = field(
        default=None,
    )

    # Battery Specific
    cycles: int | None = field(
        default=None,
    )
    state_of_charge_pct: float | None = field(
        default=None,
    )

    @staticmethod
    def array_to_external_device_list(devices: list[dict]) -> dict[str, ExternalDevice]:
        """Convert external device dict to list of ExternalDevice objects."""
        rv: dict[str, ExternalDevice] = {}

        for item in devices:
            try:
                device = ExternalDevice.from_dict(item)
            except MissingField as e:
                print(f"Error converting external device: {e}")
                continue
            rv[f"{device.type}_{device.unique_id}"] = device

        return rv

    @staticmethod
    def to_datetime(timestamp: str | int | None) -> datetime | None:
        """Convert DSRM gas-timestamp to datetime object.

        Args:
            timestamp: Timestamp string, formatted as YYMMDDHHMMSS or YYYY-MM-DDTHH:MM:SS

        Returns:
            A datetime object.
        """

        if timestamp is None:
            return None

        if isinstance(timestamp, int):
            # V1 API uses int for timestamp
            return datetime.strptime(str(timestamp), "%y%m%d%H%M%S")

        return datetime.fromisoformat(timestamp)

    @staticmethod
    def hex_to_readable(value: str | None) -> str | None:
        """Convert hex string to readable string, if possible.

        Args:
            value: String to convert, e.g. '4E47475955'

        Returns:
            A string formatted or original value when failed.
        """
        try:
            return bytes.fromhex(value).decode("utf-8")
        except (TypeError, ValueError):
            return value

    @classmethod
    # pylint: disable=too-many-statements
    def __pre_deserialize__(cls, d: dict[Any, Any]) -> dict[Any, Any]:
        _ = cls  # Unused

        if "wifi_ssid" not in d:
            # This is a v2 API response, no need to remap
            return d

        d["protocol_version"] = d.get("smr_version")
        d["meter_model"] = d.get("meter_model")
        d["unique_id"] = d.get("unique_meter_id")
        d["tariff"] = d.get("active_tariff")
        d["energy_import_kwh"] = d.get("total_power_import_kwh")
        d["energy_import_t1_kwh"] = d.get("total_power_import_t1_kwh")
        d["energy_import_t2_kwh"] = d.get("total_power_import_t2_kwh")
        d["energy_import_t3_kwh"] = d.get("total_power_import_t3_kwh")
        d["energy_import_t4_kwh"] = d.get("total_power_import_t4_kwh")
        d["energy_export_kwh"] = d.get("total_power_export_kwh")
        d["energy_export_t1_kwh"] = d.get("total_power_export_t1_kwh")
        d["energy_export_t2_kwh"] = d.get("total_power_export_t2_kwh")
        d["energy_export_t3_kwh"] = d.get("total_power_export_t3_kwh")
        d["energy_export_t4_kwh"] = d.get("total_power_export_t4_kwh")
        d["power_w"] = d.get("active_power_w")
        d["power_l1_w"] = d.get("active_power_l1_w")
        d["power_l2_w"] = d.get("active_power_l2_w")
        d["power_l3_w"] = d.get("active_power_l3_w")
        d["voltage_v"] = d.get("active_voltage_v")
        d["voltage_l1_v"] = d.get("active_voltage_l1_v")
        d["voltage_l2_v"] = d.get("active_voltage_l2_v")
        d["voltage_l3_v"] = d.get("active_voltage_l3_v")
        d["current_a"] = d.get("active_current_a")
        d["current_l1_a"] = d.get("active_current_l1_a")
        d["current_l2_a"] = d.get("active_current_l2_a")
        d["current_l3_a"] = d.get("active_current_l3_a")
        d["apparent_power_va"] = d.get("active_apparent_power_va")
        d["apparent_power_l1_va"] = d.get("active_apparent_power_l1_va")
        d["apparent_power_l2_va"] = d.get("active_apparent_power_l2_va")
        d["apparent_power_l3_va"] = d.get("active_apparent_power_l3_va")
        d["reactive_power_var"] = d.get("active_reactive_power_var")
        d["reactive_power_l1_var"] = d.get("active_reactive_power_l1_var")
        d["reactive_power_l2_var"] = d.get("active_reactive_power_l2_var")
        d["reactive_power_l3_var"] = d.get("active_reactive_power_l3_var")
        d["power_factor"] = d.get("active_power_factor")
        d["power_factor_l1"] = d.get("active_power_factor_l1")
        d["power_factor_l2"] = d.get("active_power_factor_l2")
        d["power_factor_l3"] = d.get("active_power_factor_l3")
        d["frequency_hz"] = d.get("active_frequency_hz")
        d["voltage_sag_l1_count"] = d.get("voltage_sag_l1_count")
        d["voltage_sag_l2_count"] = d.get("voltage_sag_l2_count")
        d["voltage_sag_l3_count"] = d.get("voltage_sag_l3_count")
        d["voltage_swell_l1_count"] = d.get("voltage_swell_l1_count")
        d["voltage_swell_l2_count"] = d.get("voltage_swell_l2_count")
        d["voltage_swell_l3_count"] = d.get("voltage_swell_l3_count")
        d["any_power_fail_count"] = d.get("any_power_fail_count")
        d["long_power_fail_count"] = d.get("long_power_fail_count")
        d["average_power_15m_w"] = d.get("active_power_average_w")
        d["monthly_power_peak_w"] = d.get("montly_power_peak_w")
        d["monthly_power_peak_timestamp"] = d.get("montly_power_peak_timestamp")
        d["active_liter_lpm"] = d.get("active_liter_lpm")
        d["total_liter_m3"] = d.get("total_liter_m3")
        d["external_devices"] = d.get("external_devices")

        return d


@dataclass
class ExternalDevice(BaseModel):
    """Represents externally connected device."""

    class DeviceType(Enum):
        """Device type allocations."""

        GAS_METER = "gas_meter"
        HEAT_METER = "heat_meter"
        WARM_WATER_METER = "warm_water_meter"
        WATER_METER = "water_meter"
        INLET_HEAT_METER = "inlet_heat_meter"

    unique_id: datetime = field(
        default=None, metadata={"deserialize": lambda x: Measurement.hex_to_readable(x)}
    )

    type: DeviceType | None = field(
        default=None,
        metadata={
            "deserialize": lambda x: ExternalDevice.DeviceType(x)
            if x in ExternalDevice.DeviceType.__members__.values()
            else None
        },
    )
    value: float = field(default=0.0)
    unit: str = field(default="")
    timestamp: datetime = field(
        default=None, metadata={"deserialize": lambda x: Measurement.to_datetime(x)}
    )


@dataclass(frozen=True)
class Product:
    """Represent a product."""

    model: str
    name: str | None
    url: str | None
    description: str | None

    def __str__(self):
        """Return a string representation of the product."""
        return f"HomeWizard {self.name} - {self.model}"

    def __hash__(self):
        """Hash a product for unit-test snapshots."""
        return hash((self.model, self.name, self.url, self.description))

    def __eq__(self, other):
        """Compare two products."""
        if isinstance(other, Product):
            return (
                self.model == other.model
                and self.name == other.name
                and self.url == other.url
                and self.description == other.description
            )
        return False

    @staticmethod
    def from_type(product_type: str, _: str | None = None) -> Product | None:
        """Return a Product object from a type.

        :param type: The type of the product.
        :param locale: The locale to use for the description. Currently ignored. Should be an ISO 639-1 language code.
        """

        # pylint: disable=unused-argument

        for product in PRODUCTS:
            if product.model == product_type:
                return product
        return None


PRODUCTS = {
    Product(
        "HWE-P1",
        "Wi-Fi P1 Meter",
        "https://www.homewizard.com/p1-meter/",
        "The HomeWizard P1 Meter gives you detailed insight in your electricity-, gas consumption and solar surplus.",
    ),
    Product(
        "HWE-SKT",
        "Wi-Fi Energy Socket",
        "https://www.homewizard.com/energy-socket/",
        "Measure and switch every device.",
    ),
    Product(
        "HWE-WTR",
        "Wi-Fi Watermeter",
        "https://www.homewizard.com/watermeter/",
        "Real-time water consumption insights",
    ),
    Product(
        "HWE-KWH1",
        "Wi-Fi kWh Meter 1-phase",
        "https://www.homewizard.com/kwh-meter/",
        "Measure solar panels, car chargers and more.",
    ),
    Product(
        "HWE-KWH3",
        "Wi-Fi kWh Meter 3-phase",
        "https://www.homewizard.com/kwh-meter/",
        "Measure solar panels, car chargers and more.",
    ),
    Product(
        "SDM230-wifi",
        "Wi-Fi kWh Meter 1-phase",
        "https://www.homewizard.com/kwh-meter/",
        "Measure solar panels, car chargers and more.",
    ),
    Product(
        "SDM630-wifi",
        "Wi-Fi kWh Meter 3-phase",
        "https://www.homewizard.com/kwh-meter/",
        "Measure solar panels, car chargers and more.",
    ),
    Product(
        "HWE-BAT",
        "Plug-In Battery",
        "https://www.homewizard.com/plug-in-battery/",
        "Solar energy, day and night.",
    ),
}


@dataclass(kw_only=True)
class State(BaseModel):
    """Represent current state."""

    power_on: bool | None = field(
        default=None,
    )
    switch_lock: bool | None = field(
        default=None,
    )
    brightness: int | None = field(
        default=None,
    )


@dataclass
class SystemUpdate(BaseModel):
    """Represent System update config."""

    cloud_enabled: bool = field(default=None)
    status_led_brightness_pct: int = field(default=None)
    api_v1_enabled: bool | None = field(default=None)

    def as_dict(self) -> dict[str, bool | int]:
        """Return SystemUpdate object as dict.

        Only include values that are not None.
        """
        _dict = {k: v for k, v in asdict(self).items() if v is not None}

        if not _dict:
            raise ValueError("No values to update")

        return _dict


@dataclass(kw_only=True)
class System(BaseModel):
    """Represent System config."""

    wifi_ssid: str | None = field(default=None)
    wifi_rssi_db: int | None = field(default=None)
    cloud_enabled: bool | None = field(default=None)
    uptime_s: int | None = field(default=None)
    status_led_brightness_pct: int | None = field(default=None)
    api_v1_enabled: bool | None = field(default=None)


@dataclass(kw_only=True)
class Token(BaseModel):
    """Represent Token."""

    token: str = field()
