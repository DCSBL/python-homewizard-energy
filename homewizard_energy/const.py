"""Constants for HomeWizard Energy."""

import logging

LOGGER = logging.getLogger(__name__)
