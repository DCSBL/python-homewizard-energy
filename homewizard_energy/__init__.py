"""HomeWizard Energy API library."""

from .errors import DisabledError, InvalidStateError, RequestError, UnsupportedError
from .v1 import HomeWizardEnergyV1
from .v2 import HomeWizardEnergyV2
