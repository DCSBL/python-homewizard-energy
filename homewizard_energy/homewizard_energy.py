"""HomeWizard Energy base class"""

from __future__ import annotations

import asyncio
import json
import ssl
from collections.abc import Callable, Coroutine
from http import HTTPStatus
from typing import Any, TypeVar

import async_timeout
import backoff
from aiohttp.client import (
    ClientError,
    ClientResponseError,
    ClientSession,
    ClientTimeout,
    TCPConnector,
)
from aiohttp.hdrs import METH_DELETE, METH_GET, METH_POST, METH_PUT
from mashumaro.exceptions import InvalidFieldValue, MissingField
from ..models import Device, Measurement, System, SystemUpdate, Token, State

from homewizard_energy.errors import (
    DisabledError,
    RequestError,
    ResponseError,
    UnauthorizedError,
)

class HomeWizardEnergy:
    """Base class for HomeWizard Energy API."""

    _clientsession: ClientSession | None = None
    _request_timeout: int = 10
    _host: str
    
    @property
    def host(self) -> str:
        """Return the hostname of the device.

        Returns:
            host: The used host

        """
        return self._host
    
    async def device(self) -> Device:
        """Get the device information."""
        raise NotImplementedError
    
    async def measurement(self) -> Measurement:
        """Get the current measurement."""
        raise NotImplementedError
    
    async def system(
        self,
        update: SystemUpdate | None = None,
    ) -> System:
        """Get the system information."""
        raise NotImplementedError
        
    async def state(self) -> State:
        """Get the current state."""
        raise NotImplementedError
    
    async def identify(
        self,
    ) -> None:
        """Identify the device."""
        raise NotImplementedError
    
    # async def reboot(
    #     self,
    # ) -> None:
    #     """Reboot the device."""
    #     raise NotImplementedError
