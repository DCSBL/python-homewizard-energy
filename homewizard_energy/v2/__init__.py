"""Representation of a HomeWizard Energy device."""

from __future__ import annotations

import asyncio
import logging
import ssl
from collections.abc import Coroutine
from http import HTTPStatus
from typing import Any, Callable, TypeVar

import aiohttp
import async_timeout
import backoff

from homewizard_energy.errors import (
    DisabledError,
    NotFoundError,
    RequestError,
    ResponseError,
    UnauthorizedError,
)

from .cacert import CACERT
from .models import Device, Measurement, System, SystemUpdate
from .websocket import Websocket

_LOGGER = logging.getLogger(__name__)

T = TypeVar("T")


def authorized_method(
    func: Callable[..., Coroutine[Any, Any, T]],
) -> Callable[..., Coroutine[Any, Any, T]]:
    """Decorator method to check if token is set."""

    async def wrapper(self, *args, **kwargs) -> T:
        # pylint: disable=protected-access
        if self._token is None:
            raise UnauthorizedError("Token missing")

        return await func(self, *args, **kwargs)

    return wrapper


class HomeWizardEnergyV2:
    """Communicate with a HomeWizard Energy device."""

    _clientsession: aiohttp.ClientSession | None = None
    _close_clientsession: bool = False
    _request_timeout: int = 10
    _websocket: Websocket | None = None

    def __init__(
        self,
        host: str,
        identifier: str | None = None,
        token: str | None = None,
        timeout: int = 10,
    ):
        """Create a HomeWizard Energy object.

        Args:
            host: IP or URL for device.
            id: ID for device.
            token: Token for device.
            timeout: Request timeout in seconds.
        """

        self._host = host
        self._identifier = identifier
        self._token = token
        self._request_timeout = timeout

    @property
    def host(self) -> str:
        """Return the hostname of the device.

        Returns:
            host: The used host

        """
        return self._host

    @property
    def websocket(self) -> Websocket:
        """Return the websocket object.

        Create a new websocket object if it does not exist.
        """
        if self._websocket is None:
            self._websocket = Websocket(self)
        return self._websocket

    @authorized_method
    async def device(self) -> Device:
        """Return the device object."""
        _, response = await self._request("/api")
        device = Device.from_dict(response)

        return device

    @authorized_method
    async def measurement(self) -> Measurement:
        """Return the measurement object."""
        _, response = await self._request("/api/measurement")
        measurement = Measurement.from_dict(response)

        return measurement

    @authorized_method
    async def system(
        self,
        update: SystemUpdate | None = None,
    ) -> System:
        """Return the system object."""

        if update is not None:
            data = update.as_dict()
            status, response = await self._request(
                "/api/system", method=aiohttp.hdrs.METH_PUT, data=data
            )

        else:
            status, response = await self._request("/api/system")

        if status != HTTPStatus.OK:
            error = response.get("errodr", response)
            raise RequestError(f"Failed to get system: {error}")

        system = System.from_dict(response)
        return system

    @authorized_method
    async def identify(
        self,
    ) -> bool:
        """Send identify request."""
        await self._request("/api/system/identify", method=aiohttp.hdrs.METH_PUT)
        return True

    async def get_token(
        self,
        name: str,
    ) -> str:
        """Get authorization token from device."""
        status, response = await self._request(
            "/api/user", method=aiohttp.hdrs.METH_POST, data={"name": f"local/{name}"}
        )

        if status == HTTPStatus.FORBIDDEN:
            raise DisabledError("User creation is not enabled on the device")

        if status != HTTPStatus.OK:
            error = response.get("error", response)
            raise RequestError(f"Error occurred while getting token: {error}")

        try:
            token = response["token"]
        except KeyError as ex:
            raise ResponseError("Failed to get token") from ex

        self._token = token
        return token

    @authorized_method
    async def delete_token(
        self,
        name: str | None = None,
    ) -> None:
        """Delete authorization token from device."""
        status, response = await self._request(
            "/api/user",
            method=aiohttp.hdrs.METH_DELETE,
            data={"name": name} if name is not None else None,
        )

        if status != HTTPStatus.NO_CONTENT:
            error = response.get("error", response)
            raise RequestError(f"Error occurred while getting token: {error}")

        # Our token was invalided, resetting it
        if name is None:
            self._token = None

    @property
    def token(self) -> str | None:
        """Return the token of the device.

        Returns:
            token: The used token

        """
        return self._token

    @property
    def request_timeout(self) -> int:
        """Return the request timeout of the device.

        Returns:
            request_timeout: The used request timeout

        """
        return self._request_timeout

    async def get_clientsession(self) -> aiohttp.ClientSession:
        """
        Get a clientsession that is tuned for communication with the HomeWizard Energy Device
        """

        if self._clientsession is not None:
            return self._clientsession

        def _build_ssl_context() -> ssl.SSLContext:
            context = ssl.create_default_context(cadata=CACERT)
            if self._identifier is not None:
                context.hostname_checks_common_name = True
            else:
                _LOGGER.warning("No hostname provided, skipping hostname validation")
                context.check_hostname = False  # Skip hostname validation
                context.verify_mode = ssl.CERT_REQUIRED  # Keep SSL verification active
            return context

        # Creating an SSL context has some blocking IO so need to run it in the executor
        loop = asyncio.get_running_loop()
        context = await loop.run_in_executor(None, _build_ssl_context)

        connector = aiohttp.TCPConnector(
            enable_cleanup_closed=True,
            ssl=context,
            limit_per_host=1,
        )

        self._clientsession = aiohttp.ClientSession(
            connector=connector,
            timeout=aiohttp.ClientTimeout(total=self._request_timeout),
        )

        return self._clientsession

    @backoff.on_exception(backoff.expo, RequestError, max_tries=5, logger=None)
    async def _request(
        self, path: str, method: str = aiohttp.hdrs.METH_GET, data: object = None
    ) -> Any:
        """Make a request to the API."""

        _clientsession = await self.get_clientsession()

        if _clientsession.closed:
            # Avoid runtime errors when connection is closed.
            # This solves an issue when updates were scheduled and clientsession was closed.
            return None

        # Construct request
        url = f"https://{self.host}{path}"
        headers = {
            "Content-Type": "application/json",
        }
        if self._token is not None:
            headers["Authorization"] = f"Bearer {self._token}"

        _LOGGER.debug("%s, %s, %s", method, url, data)

        try:
            async with async_timeout.timeout(self._request_timeout):
                resp = await _clientsession.request(
                    method,
                    url,
                    json=data,
                    headers=headers,
                    server_hostname=self._identifier
                    if self._identifier is not None
                    else None,
                )
                _LOGGER.debug("%s, %s", resp.status, await resp.text("utf-8"))
        except asyncio.TimeoutError as exception:
            raise RequestError(
                f"Timeout occurred while connecting to the HomeWizard Energy device at {self.host}"
            ) from exception
        except (aiohttp.ClientError, aiohttp.ClientResponseError) as exception:
            raise RequestError(
                f"Error occurred while communicating with the HomeWizard Energy device at {self.host}"
            ) from exception

        match resp.status:
            case HTTPStatus.UNAUTHORIZED:
                raise UnauthorizedError("Token rejected")
            case HTTPStatus.METHOD_NOT_ALLOWED:
                raise NotFoundError("Method not allowed")
            case HTTPStatus.NO_CONTENT:
                # No content, just return
                return (HTTPStatus.NO_CONTENT, None)
            case HTTPStatus.OK:
                pass

        content_type = resp.headers.get("Content-Type", "")
        if "application/json" in content_type:
            return (resp.status, await resp.json())

        return (resp.status, await resp.text())

    async def close(self) -> None:
        """Close client session."""
        _LOGGER.debug("Closing clientsession")
        if self._clientsession is not None:
            await self._clientsession.close()
            self._clientsession = None

    async def __aenter__(self) -> HomeWizardEnergyV2:
        """Async enter.

        Returns:
            The HomeWizardEnergyV2 object.
        """
        return self

    async def __aexit__(self, *_exc_info: Any) -> None:
        """Async exit.

        Args:
            _exc_info: Exec type.
        """
        await self.close()
