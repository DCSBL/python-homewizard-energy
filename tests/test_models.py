"""Test the helper functions."""

import pytest
from syrupy.assertion import SnapshotAssertion

from homewizard_energy.models import Product, get_verification_hostname

pytestmark = [pytest.mark.asyncio]


@pytest.mark.parametrize(
    ("model", "expected"),
    [
        ("HWE-P1", "appliance/p1dongle/1234567890"),
        ("HWE-SKT", "appliance/energysocket/1234567890"),
        ("HWE-WTR", "appliance/watermeter/1234567890"),
        ("HWE-DSP", "appliance/display/1234567890"),
        ("HWE-KWH1", "appliance/energymeter/1234567890"),
        ("SDM230-wifi", "appliance/energymeter/1234567890"),
        ("HWE-KWH3", "appliance/energymeter/1234567890"),
        ("SDM630-wifi", "appliance/energymeter/1234567890"),
        ("HWE-BAT", "appliance/battery/1234567890"),
    ],
)
async def test_get_verification_hostname_returns_expected_identifier(
    model: str, expected: str
):
    """Test if get_verification_hostname returns the expected identifier."""
    result = get_verification_hostname(model, "1234567890")
    assert result == expected


async def test_get_verification_hostname_raises_value_error():
    """Test if get_verification_hostname raises a ValueError for an unsupported model."""
    with pytest.raises(ValueError):
        get_verification_hostname("unsupported", "1234567890")


@pytest.mark.parametrize(
    ("model"),
    [
        ("HWE-P1"),
        ("HWE-SKT"),
        ("HWE-WTR"),
        ("HWE-KWH1"),
        ("HWE-KWH3"),
        ("SDM230-wifi"),
        ("SDM630-wifi"),
        ("HWE-BAT"),
    ],
)
async def test_known_products(model: str, snapshot: SnapshotAssertion):
    """Test known products."""

    product = Product.from_type(model)
    assert product is not None
    assert snapshot == product


@pytest.mark.parametrize(
    ("model"),
    [
        ("HWE-P1"),
        ("HWE-SKT"),
        ("HWE-WTR"),
        ("HWE-KWH1"),
        ("HWE-KWH3"),
        ("SDM230-wifi"),
        ("SDM630-wifi"),
        ("HWE-BAT"),
    ],
)
async def test_known_product_strings(model: str, snapshot: SnapshotAssertion):
    """Test generating product strings."""

    product = Product.from_type(model)
    assert product is not None
    assert snapshot == str(product)


@pytest.mark.parametrize(
    ("model"),
    [
        ("HWE-P2"),
        (None),
        ("WTR"),
        (""),
    ],
)
async def test_unknown_or_invalid_products(model: str, snapshot: SnapshotAssertion):
    """Test unknown or invalid products types."""

    product = Product.from_type(model)
    assert product is None
    assert snapshot == product


@pytest.mark.parametrize(
    ("lhs", "rhs", "value"),
    [
        ("HWE-P1", "HWE-P1", True),
        ("HWE-P1", "HWE-SKT", False),
        ("HWE-P1", None, False),
        ("HWE-P1", "", False),
    ],
)
async def test_comparison_between_products(lhs: str, rhs: str, value: bool):
    """Test comparison between products."""

    lhs = Product.from_type(lhs)
    rhs = Product.from_type(rhs)
    assert (lhs == rhs) == value
