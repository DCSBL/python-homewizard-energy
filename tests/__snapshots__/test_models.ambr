# serializer version: 1
# name: test_known_product_strings[HWE-BAT]
  'HomeWizard Plug-In Battery - HWE-BAT'
# ---
# name: test_known_product_strings[HWE-KWH1]
  'HomeWizard Wi-Fi kWh Meter 1-phase - HWE-KWH1'
# ---
# name: test_known_product_strings[HWE-KWH3]
  'HomeWizard Wi-Fi kWh Meter 3-phase - HWE-KWH3'
# ---
# name: test_known_product_strings[HWE-P1]
  'HomeWizard Wi-Fi P1 Meter - HWE-P1'
# ---
# name: test_known_product_strings[HWE-SKT]
  'HomeWizard Wi-Fi Energy Socket - HWE-SKT'
# ---
# name: test_known_product_strings[HWE-WTR]
  'HomeWizard Wi-Fi Watermeter - HWE-WTR'
# ---
# name: test_known_product_strings[SDM230-wifi]
  'HomeWizard Wi-Fi kWh Meter 1-phase - SDM230-wifi'
# ---
# name: test_known_product_strings[SDM630-wifi]
  'HomeWizard Wi-Fi kWh Meter 3-phase - SDM630-wifi'
# ---
# name: test_known_products[HWE-BAT]
  Product(model='HWE-BAT', name='Plug-In Battery', url='https://www.homewizard.com/plug-in-battery/', description='Solar energy, day and night.')
# ---
# name: test_known_products[HWE-KWH1]
  Product(model='HWE-KWH1', name='Wi-Fi kWh Meter 1-phase', url='https://www.homewizard.com/kwh-meter/', description='Measure solar panels, car chargers and more.')
# ---
# name: test_known_products[HWE-KWH3]
  Product(model='HWE-KWH3', name='Wi-Fi kWh Meter 3-phase', url='https://www.homewizard.com/kwh-meter/', description='Measure solar panels, car chargers and more.')
# ---
# name: test_known_products[HWE-P1]
  Product(model='HWE-P1', name='Wi-Fi P1 Meter', url='https://www.homewizard.com/p1-meter/', description='The HomeWizard P1 Meter gives you detailed insight in your electricity-, gas consumption and solar surplus.')
# ---
# name: test_known_products[HWE-SKT]
  Product(model='HWE-SKT', name='Wi-Fi Energy Socket', url='https://www.homewizard.com/energy-socket/', description='Measure and switch every device.')
# ---
# name: test_known_products[HWE-WTR]
  Product(model='HWE-WTR', name='Wi-Fi Watermeter', url='https://www.homewizard.com/watermeter/', description='Real-time water consumption insights')
# ---
# name: test_known_products[SDM230-wifi]
  Product(model='SDM230-wifi', name='Wi-Fi kWh Meter 1-phase', url='https://www.homewizard.com/kwh-meter/', description='Measure solar panels, car chargers and more.')
# ---
# name: test_known_products[SDM630-wifi]
  Product(model='SDM630-wifi', name='Wi-Fi kWh Meter 3-phase', url='https://www.homewizard.com/kwh-meter/', description='Measure solar panels, car chargers and more.')
# ---
# name: test_unknown_or_invalid_products[HWE-P2]
  None
# ---
# name: test_unknown_or_invalid_products[None]
  None
# ---
# name: test_unknown_or_invalid_products[WTR]
  None
# ---
# name: test_unknown_or_invalid_products[]
  None
# ---
